gadopt/stokes_integrators.py

```diff
@@ -169,6 +168,25 @@ def __init__(self, z, T, approximation, bcs=None, mu=1,
         # so people can overwrite parameters we've setup here
         self._solver_setup = False
 
+    def initialise(self):
```

Comment:
Maybe with @sghelichkhani 's suggestion you can avoid the need for a separate code path that only gets executed in the pure (no fs) Stokes case, i.e. you execute all the standard Stokes code and just modify afterwards what needs to be different for the fs solver.
If you do need this separate initializer:
* it should have a more descriptive name `__init__` and `setup_solver` are also initialisation - so I suggest something like `setup_equations`
* do we really need to change the API for this. The reason we defer the creation of the variational solver is that we allow the user to fiddle (add to) the equations and adjust default solver parameters. So a similar justification would be that we want to allow changing stuff that's been setup in `__init__` _before_ the equations are setup. It might well be there are/will be cases where we want to allow this but I'm not sure why we need it now: if you do need a distinct `setup_equations()` you can just call it in `StokesSolver.__init__`. When the latter is being called from `FreeSurfaceStokesSolver.__init__` it will then call your overridden  `setup_equations()` from FreeSurfaceStokesSolver. In that way the equations would already be setup in `__init__` (as it is now) so we don't have to change the API.

Reply:
Thanks Stephan, as I mentioned on Zulip I've had a go at doing this. I think it makes sense to have these separate methods because I hope this removes repeating too much code. I think this means that we don't have to repeat the looping over the terms to make sure the weak bc for the free surface normal stress gets into F. It also means we can pass the free surface id to the FreeSurfaceEquations.

Reply:
As I mentioned on Zulip I've changed this again to try and be a bit neater with the inheritance so now FreeSurfaceStokesSolver.setup_equations() calls super().setup_equations and there is also an overloaded FreeSurfaceStokesSolver.solve()